### PR TITLE
Add automated release workflows

### LINKED ISSUE #345: Improve Release Automation
## Background

Currently, the release process is outlined in [RELEASING.md](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/RELEASING.md). It is an entirely manual process, and is partially responsible for contrib releases to have fallen behind those in the core repo. To summarize the process, you must first prepare the repo for release by updating the dependency and package versions and updating the changelog, create a PR with these changes (example: #316), make additional fixes to example code/documentation in that PR, get the PR approved & merged, actually release each package to NPM, then finally publish a GitHub release.

Note: There is also a [final step](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/RELEASING.md#update-changelog) to update the changelog again, but this seemed like it was handled in #316 without the need for a separate PR.

## Proposal

Many of the commands in the release guide can be automated using GitHub Actions, but some steps like fixing breaking outdated example code are necessarily manual. GH Actions can be used in a 3-step approach to take advantage of automation and manual review:

1. An automated GH Workflow that takes a single input, the version (e.g. `0.14.0`), and produces a release PR like #316. The PR would include version bumps and the changelog update.
2. Maintainers would manually update `examples/` and `getting-started/` code as they do today, then approve the PR.
3. A second workflow would be triggered by merging of the release PR and would actually release the packages to NPM and cut the GitHub Release.

As a side note, one benefit of having this 3-step, partially manual approach is that even when new contributors are given write access so that they can be added to CODEOWNERS, they will not be able to complete a release with their access to GH workflows.

## Open Questions

* Where should the release pull request be made from? An agreed upon fork like ` dynatrace-oss-contrib`, or a branch within the main repo?
* Similar vein, who's NPM token to use for publishing? I've noticed there's [already one](https://github.com/open-telemetry/opentelemetry-js/blob/main/.github/workflows/canary.yaml#L58) in the core repo for canary tests.
* How can we enforce the second workflow only running after the release PR is merged? My current idea is to give the release PR a label like "release", and trigger the workflow when any PR is merged, but make it a no-op unless the PR has the "release" label. Open to other ideas, but GH Workflows are definitely limited here.
* Can we create a draft release in workflow 1, then publish that draft in workflow 2? Need to research GH Actions ability to do this more.

### PR DESCRIPTION
**Pull request by willarmiros**
## Which problem is this PR solving?

- Resolves #345

## Short description of the changes

- Adds 2 GitHub workflows: 1 that automatically creates a release PR, and another that publishes all packages to NPM and cuts a GH release when that PR is merged. Most details are in #345, the only still unresolved matter is whether or not to create a draft release so you can judge for yourself after seeing the video.
- You can see my runs of identical workflows on a test repo meant to simulate this repo [here](https://github.com/willarmiros/test-package-lock-repo/actions), and an example of an auto-generated release PR [here](https://github.com/willarmiros/test-package-lock-repo/pull/39). I'll also be posting a video of the whole process.

### CODE REVIEW

**Comment by willarmiros** [comment]
Darn, well I'm no good at video editing so it's too big to embed here on GitHub, but you can download the recording of the flow here: https://demo-video-otel-js-9y182zqovhunxlwv.s3-us-west-2.amazonaws.com/Screen+Recording+2021-02-25+at+8.45.39+PM.mov

**Comment by willarmiros** [comment]
@dyladan can you all take a look when you get a chance? I'm back from my vacation and will be able to address any comments.

**Comment by willarmiros** [comment]
@dyladan any chance for a review?

File: RELEASING.md
Comment (vmarchaud) [inline RELEASING.md:20]: could you use another example ? The jeager propagator is moving out to core repo (https://github.com/open-telemetry/opentelemetry-js-contrib/pull/351)

**Review by vmarchaud — APPROVED** [review]
overall LGTM, though i have a question about permission: who can launch the release workflow from github ? People with write access or do it need admin permission on the repo ?

File: RELEASING.md
Comment (willarmiros) [reply RELEASING.md]: Bad random choice :) switched to Express instrumentation

**Comment by willarmiros** [comment]
> overall LGTM, though i have a question about permission: who can launch the release workflow from github ? People with write access or do it need admin permission on the repo ?

@vmarchaud Anyone with write access can run it, but the manually-runnable workflow only creates the release PR. The PR still needs to be approved before merging, which actually triggers the publishing. So someone with write access could spam PRs (and hopefully get their write access revoked) but not actually release anything to the wild.

**Review by dyladan — APPROVED** [review]

**Comment by dyladan** [comment]
Apologize for late review. I was on vacation all last week

**Comment by dyladan** [comment]
@obecny since this is fairly important I think we should have approval from all maintainers before merging

**Review by obecny — APPROVED** [review]

**Pull request merged**
